### ISSUE TITLE
Supporting multiple brands with ES6 modules

### ISSUE THREAD

**Issue by jonathanolson** [phetsims/chipper#844]
We recommend combining all the brand code into brand/js instead of scattering to different brand/phet/js brand/phet-io/js etc.  All brands would contain the info for all brands, we would use window.phet.chipper.brand to determine what to show.  If it is important to strip out the phet-io branding or copyright text for phet sims, that can be done as a postprocessing step.  But maybe that won’t be essential.

We'll need to figure out the best way to handle phet-io and other brands, and think about how we default to adapted-from-phet (or if we should), and how other organizations would create other brands if they branch our code. Do we need to strip out resources? Do we need to strip things out for legal reasons?

Investigating...

**Comment by samreid** [comment]
@jonathanolson and I didn't get too far on this.  It seems we can combine the code in the brand repo into a single module, but for dealing with phet-io features, we brainstormed:

* dynamic imports
* symlinks
* swapping branches

We think dynamic imports is the most promising, but will need investigation.

**Comment by samreid** [comment]
What about an alternative approach where we do the following:

* When in phet-io brand, we load an additional `<script type=module>` with an entry point of `phetioEngine`.  This script is not loaded when in phet brand.
* Instead of the ifphetio! plugin, we access `phetioEngine` or other class from the namespace to get sim code => phet-io access (around 20 occurrences)
* phet-io code can still access the main code through normal import statements.

I just realized the problem with that proposal--the build process would build 2 copies of common code like axon.

How about just changing the entry point?

If it is a phet-branded build, point to `<script type="module" src="sim-main.js">`, but if it is a phet-io branded build, point to `<script type="module" src="sim-main-phetio.js">` which additionally loads the PhET-iO bundles.  We could add a build assertion to help us guarantee that no phet-io code sneaks into a phet-brand build.

**Comment by samreid** [comment]
Another strategy that simplifies our import statements: Basically, for compilation-free mode, we would always import modules from phet-io in the standard way.  import phetioEngine from '….'.  For phet-io builds, everything is fine.  For phet-brand builds, we `git checkout stubs` branch from the phet-io repo.  Or we swap paths using a webpack magic switcheroo step.

**Comment by zepumph** [comment]
@samreid, @chrisklus and I spent a fair bit of time today looking into dynamic import statements. We created small test cases and were able to get webpack to build them. They involved having two simple files, `js.js` and `js2.js` that had `export default {{your string here}}`. These test cases all worked:

<details>

```js
const hi = false;

import( hi ? './js.js' : './js2.js').then( ( { default: yes } ) => {
  console.log( yes )
} );
////////////////////////////////////////////////////////////////////////////////

const hi = false;
// ( async () => {

if ( hi ) { // like phet.chipper.brand
  import('./js2.js').then( ( { default: yes } ) => {
    console.log( yes  )
  } );
}
else {
  import('./js.js').then( ( { default: yes } ) => {
    console.log( yes + 'hi' )
  } );
}
//////////////////////////////////////////////////////////////////////////////


const hi = true;
( async () => {

  if ( hi ) {
    const yes = (await import('./js2.js')).default;
    console.log( yes );
  }
  else {
    const yes = (await import('./js.js')).default;

    console.log( yes );
  }

} )();

/////////////////////////////////////

( async () => {
  if ( window.hi ) {// with window.hi set in webpack config
    const yes = ( await import('./js2.js') ).default;
    console.log( yes );
  }
  else {
    const yes = ( await import('./js.js') ).default;

    console.log( yes );
  }

} )();
////////////////////////////////////////////////////////////////////////////////

import SimLauncher from "./SimLauncher.js";
window.hi = true;
SimLauncher.push();
( async () => {
  if ( window.hi ) {// with window.hi set in webpack config
    const yes = ( await import('./js2.js') ).default;
    console.log( yes );
  }
  else {
    const yes = ( await import('./js.js') ).default;

    console.log( yes );
  }
  SimLauncher.pop();
} )();

export default 'finished'

////////////////////////////////////////////////////////////////////////////////
//
//
function getComponent() {
  return import(/* webpackChunkName: "lodash" */ 'lodash').then( ( { default: _ } ) => {
    const element = document.createElement( 'div' );

    element.innerHTML = _.join( [ 'Hello', 'webpack' ], ' ' );

    return element;

  } ).catch( error => 'An error occurred while loading the component' );
}

getComponent().then( component => {
  document.body.appendChild( component );
} )
```
</details>

We discussed and debated briefly the pros and cons to using dynamic loading vs @samreid's proposal for stub branches above (https://github.com/phetsims/chipper/issues/844#issuecomment-580006861). A compelling argument for dynamic imports is that we have bandwidth carved out to do this right currently, and we may want to spend the time to do the more idiomatic and maintainable solution. A con of dynamic loading is that it will involve (1) polyfilling `Promise`, (2) potentially adapting many usages of phet-io types to be called from the namespace, and (3) will involve adding extra logic to `SimLauncher`. I guess by "con" there I just meant "this is what it will entail." I would like to keep brainstorming though.

**Comment by zepumph** [comment]
Oh and here is a snippet that could work in SimLauncher, basically we could make `phetioEngine` the single entry point for all things PhET-iO, and have `phetioEngine.initialize` call `phetioCommandProcessor`, `dataStream`, and (maybe) `Tandem` initializations. 

in SimLauncher maybe?
```js

  let phetioEngine = null;
  if ( phet.chipper.brand === 'phet-io' ) {// with window.hi set in webpack config
    phetioEngine = import('../../../../phet-io/js/phetioEngine.js').then( ( { default: _phetioEngine } ) => {
      phetioEngine = _phetioEngine;
      callback();
    } );
  }
  else {
    phetioEngine = ( await import('../../../../tandem/js/stub') ).default; // probably using Promise notation instead of async await
  }

**Comment by jonathanolson** [comment]
> A con of dynamic loading is that it will involve (1) polyfilling Promise

Why would this require polyfilling Promise? (And what platform supports dynamic imports but doesn't support Promises?)

**Comment by samreid** [comment]
In #846 @pixelzoom pointed out that adapted from phet and phet-io did not migrate.

**Comment by chrisklus** [comment]
@samreid and @zepumph are clear on how to proceed, untagging for developer meeting.

**Comment by zepumph** [comment]
I poked around investigating dynamic imports more, and found that it is impossible (from what I found) to build a single entry point when there are dynamic imports in that module tree. 

Thus in our local examples we were always getting more than one output js file. 

This to me doesn't automatically feel like a deal breaker, but it will add complexity to the build process depending on how we handle it. 

If everything phet-io was always loaded as one module (either phetioEngine imports all phet-io code, or something like "phetioModule" loads everything), then this may be easy to inline that source above the main output from webpack. I'm excited to discuss further with @samreid about potential options!

**Comment by zepumph** [comment]
Above is a test directory that @samreid and I worked on yesterday and today. The constraints we are shooting for are:

* conditionally omit phet-io code on build (in phet brand don't build with phet-io code)
* don't load phet-io code in uncompiled mode (phet brand)
* one output js file from the webpack step (even though this isn't totally necessary, it is simpler)
* Don't need the phet-io repos to run locally or build phet brand.

I was able to get all of those working in the above code. The key is a build-step only "ifphetio" like plugin called a "loader" which we could use to take phet-io specific modules and replace them with stubs.
https://webpack.js.org/concepts/loaders/
https://webpack.js.org/configuration/module/#ruleloaders

This is working in the above code and can be run build or not and by toggling phet/phet-io brand.

Currently main.js has one field for flipping the runtime brand, and webpack.config.js has another flag for flipping the build-time flag. (low tech!)

This feels like a solid enough proof of concept to make me confident that we are in the right direction, but I think that we should do some more research before going all in on loaders.


> one output js file from the webpack step (even though this isn't totally necessary, it is simpler)

Note that this is supported with dynamic imports through a "Magic Comment" (https://webpack.js.org/api/module-methods/#magic-comments) called `/* webpackMode: eager */`.

**Comment by samreid** [comment]
> It seems we can combine the code in the brand repo into a single module

I'm not convinced that's ideal.  That would mean always bundling the phet, adapted-from-phet and phet-io into the builds, and decoding all of those mipmaps on startup.  It would also mean the phet-io copyright text would appear in the body of the phet-io brand sim, which would not be ideal even if it is acceptable.  Maybe we should use dynamic imports for these parts too.

**Comment by jonathanolson** [comment]
Also for reference, the notes from the planning doc (so we don't lose that):

> How will we support multiple brands?
We recommend combining all the brand code into brand/js instead of scattering to different brand/phet/js brand/phet-io/js etc.  All brands would contain the info for all brands, we would use window.phet.chipper.brand to determine what to show.  If it is important to strip out the phet-io branding or copyright text for phet sims, that can be done as a postprocessing step.  But maybe that won’t be essential.
AP: Would CM be a better fit than MK?
CM: I feel like I am a few weeks behind on Natural Selection and pH scale PhET-iO instrumentation.
MK: SR and I investigated this for about 3 hours yesterday, and it seems to require deeper knowledge about phet-io and how it is used in Sim.

**Comment by zepumph** [comment]
In https://github.com/phetsims/phet-io/issues/1619 and https://github.com/phetsims/phet-io/issues/1618 I readied code to try to get rid of the ifphetio! plugin, and limit the entry points into phet-io specific code.

Basically now there is just phetioEngine imported in SimLauncher, and phetioEngine imports everything else, dataStream/phetioCommandProcessor/PhetioStateEngine etc. 

I think we are in a good place to start prototyping in es6 modules for a way to include phet-io code dynamically.

**Comment by zepumph** [comment]
For this issue specifically, @samreid and I met and were able to learn a bit more about eslint and dynamic imports. We found that using the `babel-eslint` parser allows dynamic imports, but we can't use it everywhere because when testing that, lint-everything took over 10 minutes (then I aborted). @samreid said his took 15.

We can still have this work though if we use the `babel-eslint` parser sparingly with custom `.eslintrc` files like where SimLauncher lives. It could be something like:

```
  parser: "babel-eslint",
  "parserOptions": {
    "sourceType": "module",
    "allowImportExportEverywhere": true
  },
```

It is still unclear though if we need to npm install `babel-eslint` in every repo that would lint joist, or just in chipper (ideal), or perhaps npm install in joist (since that is where the parser config file lives). I'll investigate more in the future.

**Comment by zepumph** [comment]
Correction: Over in https://github.com/phetsims/chipper/issues/857 @samreid said it took 6 minutes.

**Comment by samreid** [comment]
Here's a migrated SimLauncher that uses dynamic imports and waits until it's ready:

<details>

```js
// Copyright 2013-2020, University of Colorado Boulder
/**
 * Launches a PhET Simulation, after preloading the specified images.
 *
 * @author Sam Reid (PhET Interactive Simulations)
 */

import Random from '../../dot/js/Random.js';
import checkNamespaces from '../../joist/js/checkNamespaces.js';
import joist from '../../joist/js/joist.js';
import Property from '../../axon/js/Property';

const phetioEngineProperty = new Property( null );

if ( phet.chipper.brand === 'phet-io' ) {

  import( /* webpackMode: "eager" */ '../../phet-io/js/phetioEngine').then( module => {
    phetioEngineProperty.value = module.default;
  } );
}

const SimLauncher = {

  /**
   * Launch the Sim by preloading the images and calling the callback.
   *
   * @param callback the callback function which should create and start the sim, given that the images are loaded
   * @public - to be called by main()s everywhere
   */
  launch: function( callback ) {
    assert && assert( !window.phet.launchCalled, 'Tried to launch twice' );

    //Signify that the SimLauncher was called, see https://github.com/phetsims/joist/issues/142
    window.phet.joist.launchCalled = true;

    // image elements to remove once we are fully loaded
    const elementsToRemove = [];

    let waitingForImages = true;

    function doneLoadingImages() {
      waitingForImages = false;
      proceedIfReady();
    }

    function proceedIfReady() {
      if ( waitingForImages || phetioEngineProperty.value === null ) {
        return;
      }
      window.phet.joist.launchSimulation = function() {

        // once launchSimulation has been called, the wrapper is ready to receive messages because any listeners it
        // wants have been set up by now.
        if ( phet.phetio ) {
          phetioEngineProperty.value.onCrossFrameListenersReady();
        }

        // Provide a global Random that is easy to use and seedable from phet-io for playback
        // phet-io configuration happens after SimLauncher.launch is called and before phet.joist.launchSimulation is called
        phet.joist.random = new Random( { staticSeed: true } );

        // Instantiate the sim and show it.
        callback();
      };

      // PhET-iO simulations support an initialization phase (before the sim launches)
      if ( phet.phetio ) {
        phetioEngineProperty.value.initialize(); // calls back to window.phet.joist.launchSimulation
      }

      if ( phet.chipper.queryParameters.postMessageOnReady ) {
        window.parent && window.parent.postMessage( JSON.stringify( {
          type: 'ready',
          url: window.location.href
        } ), '*' );
      }

      if ( ( phet.phetio && !phet.phetio.queryParameters.phetioStandalone ) ||
           phet.chipper.queryParameters.playbackMode ) {

        // Wait for phet-io to finish adding listeners. It will direct the launch from there.
      }
      else {
        window.phet.joist.launchSimulation();
      }

    }

    phetioEngineProperty.link( proceedIfReady );

    // if image dimensions exist, immediately fire the "all images loaded" event
    let loaded = 0;

    // Taken from http://stackoverflow.com/questions/1977871/check-if-an-image-is-loaded-no-errors-in-javascript
    function isImageOK( img ) {

      // During the onload event, IE correctly identifies any images that
      // weren't downloaded as not complete. Others should too. Gecko-based
      // browsers act like NS4 in that they report this incorrectly.
      if ( !img.complete ) {
        return false;
      }

      // However, they do have two very useful properties: naturalWidth and
      // naturalHeight. These give the true size of the image. If it failed
      // to load, either of these should be zero.
      if ( typeof img.naturalWidth !== 'undefined' && img.naturalWidth === 0 ) {
        return false;
      }

      // No other way of checking: assume it’s ok.
      return true;
    }

    //For the images that were written to base64 format using requirejs, make sure they are loaded.
    //img.src = base64 is asynchronous on IE10 and OSX/Safari, so we have to make sure they loaded before returning.
    if ( window.phetImages ) {
      for ( let i = 0; i < window.phetImages.length; i++ ) {
        const phetImage = window.phetImages[ i ];

        // For built versions that use phet-io, the simulation may have already loaded all of the images, so
        // check them here before scheduling them for load.
        if ( isImageOK( phetImage ) ) {
          loaded++;
          if ( loaded === window.phetImages.length ) {
            doneLoadingImages();
          }
        }
        else {
          phetImage.onload = function() {
            loaded++;
            if ( loaded === window.phetImages.length ) {
              doneLoadingImages();
            }
          };
        }

      }
    }
    else {
      doneLoadingImages();
    }

    $( window ).load( function() {
      // if images were not loaded immediately, signal the "all images loaded" event

      // we wait for here to remove the images from the DOM, otherwise IE9/10 treat the images as completely blank!
      _.each( elementsToRemove, function( element ) {

        //TODO: Why is this null sometimes?  see https://github.com/phetsims/joist/issues/388
        if ( element.parentNode ) {
          element.parentNode.removeChild( element );
        }
      } );
    } );

    // Check namespaces if assertions are enabled, see https://github.com/phetsims/joist/issues/307.
    assert && checkNamespaces();
  }
};

joist.register( 'SimLauncher', SimLauncher );

export default SimLauncher;
```

</details>

To get this to work in Gravity and Orbits, I had to move creation of an instrumented object downstream to after the sim started:

<details>

```js
// Copyright 2013-2019, University of Colorado Boulder

/**
 * Main entry point for the sim.
 *
 * @author Aaron Davis (PhET Interactive Simulations)
 */

import GravityAndOrbitsColorProfile from '../../gravity-and-orbits/js/common/GravityAndOrbitsColorProfile.js';
import GlobalOptionsNode from '../../gravity-and-orbits/js/common/view/GlobalOptionsNode.js';
import gravityAndOrbitsStrings from '../../gravity-and-orbits/js/gravity-and-orbits-strings.js';
import ModelScreen from '../../gravity-and-orbits/js/model/ModelScreen.js';
import ToScaleScreen from '../../gravity-and-orbits/js/toScale/ToScaleScreen.js';
import Sim from '../../joist/js/Sim.js';
import SimLauncher from '../../joist/js/SimLauncher.js';
import Tandem from '../../tandem/js/Tandem.js';

const gravityAndOrbitsTitleString = gravityAndOrbitsStrings[ 'gravity-and-orbits' ].title;


const simOptions = {
  credits: {
    leadDesign: 'Emily B. Moore, Noah Podolefsky, Amy Rouinfar',
    softwareDevelopment: 'Aaron Davis, Jesse Greenberg, Jon Olson, Sam Reid',
    team: 'Trish Loeblein, Ariel Paul, Kathy Perkins',
    qualityAssurance: 'Steele Dalton, Ethan Johnson, Elise Morgan, Oliver Orejola, Ben Roberts, Bryan Yoelin',
    thanks: 'Thanks to Mobile Learner Labs for working with the PhET development team to convert this simulation to ' +
            'HTML5.'
  }
};

SimLauncher.launch( () => {

  // Eagerly create GlobalOptionsNode so it works smoothly with PhET-iO
  const x = new GlobalOptionsNode( Tandem.ROOT.createTandem( 'global' ).createTandem( 'view' ).createTandem( 'globalOptionsNode' ) )
  simOptions.createOptionsDialogContent = ()=>x;

  const modelScreen = new ModelScreen( {
    backgroundColorProperty: GravityAndOrbitsColorProfile.backgroundProperty,
    tandem: Tandem.ROOT.createTandem( 'modelScreen' )
  } );
  const toScaleScreen = new ToScaleScreen( {
    backgroundColorProperty: GravityAndOrbitsColorProfile.backgroundProperty,
    tandem: Tandem.ROOT.createTandem( 'toScaleScreen' )
  } );
  new Sim( gravityAndOrbitsTitleString, [ modelScreen, toScaleScreen ], simOptions ).start();
} );
```

</details>

After these changes, the sim launches in standalone mode, datastream mode and runs in studio, with the following problems and caveats:
1. I had to disable assertions because it says `Property` is getting namespaced twice.
2. The wrong logo is being displayed
3. In studio, there is parse error on events from postMessage which I don't understand.

**Comment by samreid** [comment]
Oh my, WebStorm "auto import" did not include the suffix, so the import statement was:

```js
import Property from '../../axon/js/Property.js';
```

This triggered it to load as a different file, when I changed the import to:
```js
import Property from '../../axon/js/Property.js';
```

It worked correctly.  I think we discovered in another issue that Windows machines require the suffix, so hopefully we can get Webstorm to auto-import and include the file suffix.

**Comment by samreid** [comment]
Why is this code in `Brand`?

```js
isPhetApp: phet.chipper.queryParameters[ 'phet-app' ] || phet.chipper.queryParameters[ 'phet-android-app' ],
```

**Comment by samreid** [comment]
I have significant working copy changes for this, please communicate with me if you want to proceed and I haven't committed yet.

**Comment by samreid** [comment]
It seems the babel-eslint parser needs to be specified adjacent to where the npm install happens, so in the package.json:

```json
  "eslintConfig": {
    "extends": "../chipper/eslint/sim_eslintrc.js",
    "parser": "babel-eslint",
    "parserOptions": {
      "sourceType": "module",
      "allowImportExportEverywhere": true
    }
  }
```

I'll need to check whether we need one or both of these in joist/package.json.

```json
    "babel-eslint": "^10.0.3",
    "eslint": "^6.8.0"
```

**Comment by samreid** [comment]
The reference at `https://github.com/babel/babel-eslint/issues/788#issuecomment-571925824` shows how to use a path for parser.

**Comment by samreid** [comment]
I messaged on Slack:

> After commits in https://github.com/phetsims/chipper/issues/844, a new devDependency has been added to chipper, please `npm update` or `npm install` before your next `lint`

**Comment by samreid** [comment]
When I used the stub, it looks like webpack put in 2 copies of it.  Will it put 2 copies of phetioEngine as well?

**Comment by samreid** [comment]
After the commit, phet build uses the stub and contains no phet-io code.  phet-io brand contains the phet-io code, and both run nicely.  However, when deleting the phet-io repo, and running `grunt --brands=phet`, there is an error:

```
    errors: [
      ModuleNotFoundError: Module not found: Error: Can't resolve '../../phet-io/js/phetioEngine.js' in '/Users/samreid/github/joist/js'
        at /Users/samreid/github/chipper/node_modules/webpack/lib/Compilation.js:925:10
        at /Users/samreid/github/chipper/node_modules/webpack/lib/NormalModuleFactory.js:401:22
        at /Users/samreid/github/chipper/node_modules/webpack/lib/NormalModuleFactory.js:130:21
        at /Users/samreid/github/chipper/node_modules/webpack/lib/NormalModuleFactory.js:224:22
```

**Comment by samreid** [comment]
Working nicely after the commit!  Whew.

**Comment by samreid** [comment]
Some changes in master warrant review before publication, particularly SimLauncher and its "dynamic modules".

**Comment by samreid** [comment]
I still have some unpushed changes.  Things are going in the direction but now there's a problem that the string aggregator doesn't know how to traverse dynamic imports, so `getLinks` is failing.  We should just move that out of the dynamic brands probably.

**Comment by samreid** [comment]
These changes are failing on CT with this error:

<details>

```
acid-base-solutions : build : phet
Failure to grunt snapshot-1581787269140/acid-base-solutions:
Running "lint-all" task
Fatal error: Perennial task failed:
Error: Failed to load parser '../chipper/node_modules/babel-eslint' declared in 'joist/package.json': Cannot find module '../chipper/node_modules/babel-eslint'
Require stack:
- /data/share/phet/continuous-testing/snapshot-1581787269140/joist/package.json
  at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
  at Function.resolve (internal/modules/cjs/helpers.js:33:19)
  at Object.resolve (/data/share/phet/continuous-testing/snapshot-1581787269140/chipper/node_modules/eslint/lib/shared/relative-module-resolver.js:44:50)
  at ConfigArrayFactory._loadParser (/data/share/phet/continuous-testing/snapshot-1581787269140/chipper/node_modules/eslint/lib/cli-engine/config-array-factory.js:870:45)
  at ConfigArrayFactory._normalizeObjectConfigDataBody (/data/share/phet/continuous-testing/snapshot-1581787269140/chipper/node_modules/eslint/lib/cli-engine/config-array-factory.js:665:32)
  at _normalizeObjectConfigDataBody.next (<anonymous>:null:null)
  at ConfigArrayFactory._normalizeObjectConfigData (/data/share/phet/continuous-testing/snapshot-1581787269140/chipper/node_modules/eslint/lib/cli-engine/config-array-factory.js:596:20)
  at _normalizeObjectConfigData.next (<anonymous>:null:null)
  at createConfigArray (/data/share/phet/continuous-testing/snapshot-1581787269140/chipper/node_modules/eslint/lib/cli-engine/config-array-factory.js:340:25)
  at ConfigArrayFactory.loadInDirectory (/data/share/phet/continuous-testing/snapshot-1581787269140/chipper/node_modules/eslint/lib/cli-engine/config-array-factory.js:433:16)
  at CascadingConfigArrayFactory._loadConfigInAncestors (/data/share/phet/continuous-testing/snapshot-1581787269140/chipper/node_modules/eslint/lib/cli-engine/cascading-config-array-factory.js:328:46)
  at CascadingConfigArrayFactory.getConfigArrayForFile (/data/share/phet/continuous-testing/snapshot-1581787269140/chipper/node_modules/eslint/lib/cli-engine/cascading-config-array-factory.js:272:18)
  at FileEnumerator._iterateFilesRecursive (/data/share/phet/continuous-testing/snapshot-1581787269140/chipper/node_modules/eslint/lib/cli-engine/file-enumerator.js:431:49)
  at _iterateFilesRecursive.next (<anonymous>:null:null)
  at FileEnumerator.iterateFiles (/data/share/phet/continuous-testing/snapshot-1581787269140/chipper/node_modules/eslint/lib/cli-engine/file-enumerator.js:251:49)
  at iterateFiles.next (<anonymous>:null:null)
  at CLIEngine.executeOnFiles (/data/share/phet/continuous-testing/snapshot-1581787269140/chipper/node_modules/eslint/lib/cli-engine/cli-engine.js:761:48)
  at module.exports (/data/share/phet/continuous-testing/snapshot-1581787269140/chipper/js/grunt/lint.js:91:22)
  at grunt.registerTask.wrapTask (/data/share/phet/continuous-testing/snapshot-1581787269140/chipper/js/grunt/Gruntfile.js:228:5)
  at Object.<anonymous> (/data/share/phet/continuous-testing/snapshot-1581787269140/chipper/js/grunt/Gruntfile.js:93:13)
  at Object.thisTask.fn (/data/share/phet/continuous-testing/snapshot-1581787269140/acid-base-solutions/node_modules/grunt/lib/grunt/task.js:70:16)
  at Object.<anonymous> (/data/share/phet/continuous-testing/snapshot-1581787269140/acid-base-solutions/node_modules/grunt/lib/util/task.js:294:30)
  at Task.runTaskFn (/data/share/phet/continuous-testing/snapshot-1581787269140/acid-base-solutions/node_modules/grunt/lib/util/task.js:244:24)
  at Task.<anonymous> (/data/share/phet/continuous-testing/snapshot-1581787269140/acid-base-solutions/node_modules/grunt/lib/util/task.js:293:12)
  at /data/share/phet/continuous-testing/snapshot-1581787269140/acid-base-solutions/node_modules/grunt/lib/util/task.js:220:11
  at process._tickCallback (internal/process/next_tick.js:61:11)
  at Function.Module.runMain (internal/modules/cjs/loader.js:834:11)
  at startup (internal/bootstrap/node.js:283:19)
  at bootstrapNodeJSCore (internal/bootstrap/node.js:622:3)

Full Error details:
{
  "code": "MODULE_NOT_FOUND"
}


Approximately 2/15/2020, 10:21:09 AM
```

</details>

On slack:

1:32
Does CT not auto npm update in chipper?






Jonathan Olson  2:04 PM
I thought it did?

Sam Reid  3:17 PM
I thought so too, maybe something else is happening

**Comment by samreid** [comment]
I can reproduce that problem if I `cd chipper` and `rm -rf node_modules` then delete `"babel-eslint": "^10.0.3",` and `npm update`.  Next we should check if this failed to npm update on bayes.

**Comment by samreid** [comment]
Looks like I neglected to push a commit, It’s pushed now and I’ll keep an eye on the next columns.

**Comment by samreid** [comment]
After commits here and in #879, strings are fixed again and the correct text shows in the About dialog based on the brand.  Sims are running in compilation-free and built mode for master, and after migration.  If I delete the `phet-io` repo, I can still launch the sim in compilation free mode and still build the sim and it runs OK.  I do get this warning during the build though:

```
>> WARNING404: Skipping potentially non-public dependency phet-io
```

Next step is to get the logo in the about dialog to display correctly.  At the moment, it is always the phet brand logo.

**Comment by samreid** [comment]
By adding `logo` to `Brand`, there is a problem where there are 25 `phetImages` on startup, then the 26th sneaks in later, after the initial `25` are listened for.  So we have another problem (like the joist strings) where loading dynamically will cause extra complication.  We could update to a pattern where instead of pushing an image into the `phetioImages` array, we immediately add the listener and schedule that as work for SimLauncher.  This would take some reworking but should be OK, as long as the image is scheduled for loading *before* the dynamic module signifies that it has been fully loaded.  Not sure if there is a better way.

**Comment by samreid** [comment]
Is `elementsToRemove` unused?

**Comment by zepumph** [comment]
I looked at SimLauncher briefly and things look really great! Thanks for the hard work on this. Do you want to pair to co-review and go over loose ends sometime this week?

**Comment by chrisklus** [comment]
I've pulled all and am getting the following error when trying to run energy-forms-and-changes-locally:

```
assert.js?bust=1582317362116:22 Uncaught Error: Assertion failed: Sims usually have images
    at window.assertions.assertFunction (assert.js?bust=1582317362116:22)
    at SimLauncher.js?bust=1582317362244:161
    at Property.link (Property.js?bust=1582317362244:360)
    at Object.launch (SimLauncher.js?bust=1582317362244:157)
    at energy-forms-and-changes-main.js?bust=1582317362244:37
    at Object.execCb (require-2.3.6.js?bust=1582317362116:1696)
    at Module.check (require-2.3.6.js?bust=1582317362116:883)
    at Module.<anonymous> (require-2.3.6.js?bust=1582317362116:1139)
    at require-2.3.6.js?bust=1582317362116:134
    at require-2.3.6.js?bust=1582317362116:1189
```

**Comment by zepumph** [comment]
We had a commit from chipper that was not yet pushed, sorry!

**Comment by chrisklus** [comment]
Working now, thanks!

**Comment by zepumph** [comment]
@samreid and I worked out loose ends here, and created a side issue for after migration. All brands are working correctly, and we confirmed that phet-io code is not added in phet brand builds, and a lack of a phet-io repo won't break a phet build. Closing.
